lib/cisco_node_utils/yang.rb

```diff
@@ -0,0 +1,156 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'json'
+
+module Cisco
+  # class Cisco::Yang
+  # Utility class mainly containing methods to compare YANG configurations
+  # in order to determine if the running config is in-sync with the
+  # desired config.
+  class Yang
+    # Is the specified yang string empty?
+    def self.empty?(yang)
+      !yang || yang.empty?
+    end
+
+    # Given a current and target YANG configuration, returns true if
+    # the configurations are in-sync, relative to a "merge_config" action
+    def self.insync_for_merge?(target, current)
+      target_hash = self.empty?(target) ? {} : JSON.parse(target)
+      current_hash = self.empty?(current) ? {} : JSON.parse(current)
+
+      !needs_something?(:merge, target_hash, current_hash)
+    end
+
+    # Given a current and target YANG configuration, returns true if
+    # the configuration are in-sync, relative to a "replace_config" action
+    def self.insync_for_replace?(target, current)
+      target_hash = self.empty?(target) ? {} : JSON.parse(target)
+      current_hash = self.empty?(current) ? {} : JSON.parse(current)
+
+      !needs_something?(:replace, target_hash, current_hash)
+    end
+
+    # usage:
+    #   needs_something?(op, target, run)
+    #
+    #   op     - symbol - If value is not :replace, it's assumed to be :merge.
+    #                     Indicates to the function whether to check for a
+    #                     possible merge vs. replace
+    #
+    #   target - JSON   - JSON tree representing target configuration
+    #
+    #   run    - JSON   - JSON tree representing target configuration
+    #
+    #
+    # Needs merge will determine if target and run differ
+    # sufficiently to necessitate running the merge command.
+    #
+    # The logic here amounts to determining if target is a subtree
+    # of run, with a tiny bit of domain trickiness surrounding
+    # elements that are arrays that contain a single nil element
+    # that is required for "creating" certain configuration elements.
+    #
+    # There are ultimately 3 different types of elements in a json
+    # tree.  Hashes, Arrays, and leaves.  While hashes and array values
+    # are organized with an order, the logic here ignores the order.
+    # In fact, it specifically attempts to match assuming order
+    # doesn't matter.  This is largely to allow users some freedom
+    # in specifying the config in the manifest.  The gRPC interface
```

Comment:
Should you reference manifest in this context?  This could in theory be used by any orchestration frame work right, not just puppet?